### PR TITLE
[Basic] Add FileManager and start caching `stat` calls

### PR DESCRIPTION
**Pull request by harlanhaskins**
This is both an optimization -- we avoid `stat`-ing multiple time -- and a mitigation for a correctness issue. We want to keep track of the file stats from the moment we open a file, and we don't want to stat it after we've opened & read it. This avoids racing where a file might change during compilation, and we record the new modification time when we've read the old one.

### CODE REVIEW

File: include/swift/Basic/FileManager.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+/// A file manager that caches file stats, both to avoid re-statting files, but
+/// also to ensure we don't accidentally record newer stats than when we first
+/// read the file.
+class FileManager: public llvm::RefCountedBase<FileManager> {
+private:
+  friend class SourceManager;
+  clang::MemorizeStatCalls cachedStats;
+  llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fileSystem;
+  StringRef getCachedFilename(StringRef path);
```
Comment (benlangmuir) [inline include/swift/Basic/FileManager.h]: The values returned here are invalidated if anyone calls `setFileSystem`.  It looks like this is used by the source manager.  Assuming this is actually okay in practice, we should document this pitfall.

**Comment by harlanhaskins** [comment]
This can't be tested until https://reviews.llvm.org/D60735 goes in.

**Comment by harlanhaskins** [comment]
https://github.com/apple/swift-clang/pull/300

@swift-ci please test

**Comment by harlanhaskins** [comment]
apple/swift-clang#300

@swift-ci please test

**Comment by harlanhaskins** [comment]
apple/swift-clang#300

@swift-ci please test Linux

**Comment by harlanhaskins** [comment]
Closing for now.

**Pull request closed without merging**